### PR TITLE
Add amazon ATM support

### PR DESCRIPTION
**Pull request by evgenizhuk**

### CODE REVIEW

**Comment by evgenizhuk** [comment]
This PR adds amazon transparent market place support (https://aps.amazon.com/aps/transparent-ad-marketplace/)

**Comment by scott-thrillist** [comment]
@evgenizhuk @jaanauati what are you thoughts on keeping this component brand agnostic of add-ons? We use Amazon TAM + react-dfp as well but simply utilize the built-in AdSlot props to connect it all and it works fine.

Just bringing this up because there are other bidding systems out there and if we go down the route of internally supporting other systems this project could balloon in scope & size. Open to discussion here.

**Comment by scott-thrillist** [comment]
Ps- here's the original PR that allowed react-dfp to work with Amazon TAM

https://github.com/jaanauati/react-dfp/pull/158

**Comment by yahenizhuk** [comment]
> Ps- here's the original PR that allowed react-dfp to work with Amazon TAM
> 
> #158

@scott-thrillist thank you for the reference, I've missed this prop. Probably, it makes sense to update documentation.

**Comment by yahenizhuk** [comment]
@scott-thrillist could you provide an example? As I understand for amazon TAM enabling `DFPSlotsProvider:autoReload:adUnit` should be set to `true` and register bids fetching as `onSlotRegister` callback. Does a DFP single request architecture mode work correctly with it? Thank you

**Comment by scott-thrillist** [comment]
@yahenizhuk Here's a snippet I created for someone else. I may clean this up and make an example within this project when some time opens up.

https://gist.github.com/scott-thrillist/bf7d429f6c78a417fe4717a2e1791dc1

A potential downside here is we're not making a single request to Amazon but ideally you should be lazy loading ads to increase your viewability metrics. This would make new, separate requests as you scroll down the page vs. a single bigger payload and longer wait times during load.

**Comment by yahenizhuk** [comment]
@scott-thrillist Thank you for the example. Unfortunately, it doesn't fit our requirements. I'll try to approach it with a plugin.

**Comment by scott-thrillist** [comment]
@yahenizhuk Gotcha and thank you. Can I ask what are the requirements that this doesn't meet? Always interested in understanding how other organizations implement ad bidding with DFP.

**Comment by yahenizhuk** [comment]
Hi @scott-thrillist , thank you for quick answers. Unfortunately, I don't think I can share a lot due to our policy. I can say only that SRA mode is critically important for us also we need to add bidding, not to all ad units on the page so with `disableInitialLoad` in provider and refreshes in adSlot it's hard to control ads behavior.
But I like your idea about the plugin. I will continue with this and will return with the PR when it will be done. I think this PR can be closed. Thank you!

**Comment by scott-thrillist** [comment]
Thanks for sharing. We implement conditional ad bidding as well, and simply refresh the ad sans-TAM if it doesn't meet the condition. But I understand everyone has different requirements. Happy coding

File: js/manager.js [hunk 1]
```diff
@@ -407,6 +429,22 @@ const DFPManager = Object.assign(new EventEmitter().setMaxListeners(0), {
     }
   },
 
+  registerAmazonSlot(slotID, sizes, adUnit) {
```
Comment (jaanauati) [inline js/manager.js]: I wonder if we should move/put most of these new code/logic into a separated js, example: amazonManager.js, this way the code would be easier to maintain/extend in the short/long term, example:
amazonManager.js
 + registerSlot()  (registerAmazonSlot())
 + getSlots()  (getAmazonSlots())
 + count()  (isAmazonSlotsEmpty()).

what do you think guys?

File: js/adslot.js [hunk 2]
```diff
@@ -103,6 +105,13 @@ export class AdSlot extends React.Component {
       ...this.state,
       slotShouldRefresh: this.slotShouldRefresh,
     });
+    if (this.props.withAmazon === true) {
+      DFPManager.registerAmazonSlot(
```
[the remark this answers is not in the document]
Comment (evgenizhuk) [reply js/adslot.js]: Sure, good point, thanks

File: js/manager.js (hunk 1 above)
Comment (evgenizhuk) [reply js/manager.js]: Absolutely agree, will move to a separate module

**Pull request closed without merging**